pkg/lifecycle/service.go

```diff
@@ -176,6 +185,77 @@ func (s *Service) Start(
 	return nil
 }
 
+// Restart stops an existing pipeline and replaces their nodes.
+func (s *Service) Restart(ctx context.Context, rp *runnablePipeline) error {
+	rp, ok := s.runningPipelines.Get(rp.pipeline.ID)
+	if !ok {
+		return cerrors.Errorf("pipeline %s can't be restarted: %w", rp.pipeline.ID, pipeline.ErrInstanceNotFound)
+	}
+
+	// In case we want to restart a running pipeline.
+	if rp.pipeline.GetStatus() == pipeline.StatusRunning {
+		if err := s.Stop(ctx, rp.pipeline.ID, true); err != nil {
+			return cerrors.Errorf("could not stop pipeline %s: %w", rp.pipeline.ID, err)
+		}
+	}
+
+	s.logger.Debug(ctx).Str(log.PipelineIDField, rp.pipeline.ID).Msg("restarting pipeline")
+	s.logger.Trace(ctx).Str(log.PipelineIDField, rp.pipeline.ID).Msg("swapping nodes")
+
+	nodes, err := s.buildNodes(ctx, rp.pipeline)
+	if err != nil {
+		return cerrors.Errorf("could not build new nodes for pipeline %s: %w", rp.pipeline.ID, err)
+	}
+
+	rp.n = nodes
+
+	s.logger.Trace(ctx).Str(log.PipelineIDField, rp.pipeline.ID).Msg("running nodes")
+	if err := s.runPipeline(ctx, rp); err != nil {
+		return cerrors.Errorf("failed to run pipeline %s: %w", rp.pipeline.ID, err)
+	}
+	s.logger.Info(ctx).Str(log.PipelineIDField, rp.pipeline.ID).Msg("pipeline re-started ")
+
+	return nil
+}
+
+// RestartWithBackoff restarts a pipeline with a backoff.
+func (s *Service) RestartWithBackoff(ctx context.Context, rp *runnablePipeline) error {
+	// backoffCfg.Attempt() returns a float64
+	attempt := int(rp.backoffCfg.Attempt())
+	duration := rp.backoffCfg.Duration()
+
+	if attempt > s.errRecoveryCfg.MaxRetries {
+		return cerrors.FatalError(cerrors.Errorf("failed to recover pipeline %s after %d attempts: %w", rp.pipeline.ID, attempt, pipeline.ErrPipelineCannotRecover))
+	}
+
+	// This results in a default delay progression of 1s, 2s, 4s, 8s, 16s, [...], 10m, 10m,... balancing the need for recovery time and minimizing downtime.
+	timer := time.NewTimer(duration)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-time.After(duration):
+		<-timer.C
+	}
+
+	// Get status of pipeline, if it recovers (it's running),
+	pl, err := s.pipelines.Get(ctx, rp.pipeline.ID)
```

Comment:
Btw, `rp.pipeline` is already the pipeline, no need to fetch it again through the service 😉